### PR TITLE
[FIRRTL] Split name preservation semantics into a dedicated attribute

### PR DESCRIPTION
**Pull request by uenoku**
This PR adds BoolAttr to explicitly represent name preservations of node, wire and register ops. Previously, `name` is directly used to represent the preservation kind (useless or interesting name) but it would be nicer to have the information explicitly, considering that name attributes are used as SSA names too. 

Example:
```scala
circuit Foo:
  module Foo:
    input a: UInt<1>
    node b = a
    node _c = a
```  

Parse only output:
```mlir
 firrtl.module @Foo(in %a: !firrtl.uint<1>) {
      %b = firrtl.node  %a  : !firrtl.uint<1>
      // same as %b = firrtl.node interesting_name %a: !firrtl.uint<1>
      %_c = firrtl.node droppable_name  %a  : !firrtl.uint<1>
 }
```

### CODE REVIEW

**Comment by uenoku** [comment]
I'd like to use the terminology "droppable name" instead of "useless name" to  distinguish them clearly. 

* "useless name" is a name that is empty or starts with "_". Otherwise, the name is "interesting name". 
* "droppable name" is determined by a boolean flag owned by operations.

File: include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td [hunk 1]
```diff
@@ -125,34 +125,34 @@ def GetWidthAsIntAttr : NativeCodeCall<
 
 // node(x,name),!dnt -> node(x, "")
 def DropNameNode : Pat<
-  (NodeOp $x, $name, $annotations, $inner_sym),
-  (NodeOp $x, (GetEmptyString), $annotations, $inner_sym),
+  (NodeOp $x, $name, $annotations, $inner_sym, $hasDroppableName),
+  (NodeOp $x, (GetEmptyString),  $annotations, $inner_sym, $hasDroppableName),
```
Comment (mikeurbach) [inline include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td:129]: For these "DropName" transformations (didn't even know they existed!), should we be setting hasDroppableName to `true` rather than carrying over what it was previously? It seems like we could potentially have an empty name string with hasDroppableName=true.

File: include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td [hunk 2]
```diff
@@ -98,7 +98,7 @@ def InstanceOp : ReferableDeclOp<"instance", [HasParent<"firrtl::FModuleOp, firr
   }];
 }
 
-def MemOp : ReferableDeclOp<"mem"> {
+def MemOp : ReferableDeclOp<"mem", [FNamableOp]> {
```
Comment (mikeurbach) [inline include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td:101]: Should FNamableOp just be added to the base ReferableDeclOp? I see it isn't on InstanceOp now, but maybe it could be there too?

**Review by mikeurbach — COMMENTED** [review]
This looks pretty good to me, just a couple questions.

Also- I generally have this question: if you need a boolean now, should it just be an enum? Sometimes you need two options, but soon you might need three, and it's easier to add a third if you don't change the type later. Anyway, that might be premature optimization since I don't know a concrete third option right now.

File: include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td (hunk 2 above)
Comment (uenoku) [reply include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td:101]: Good point. I intentionally avoided InstanceOp because the dropname canonicalizer is not implemented for InstanceOp name. But on the second thought, there should be no reason to exclude InstanceOp so I'll move FNamebleOp into ReferrableOp.

**Comment by uenoku** [comment]
> Also- I generally have this question: if you need a boolean now, should it just be an enum? Sometimes you need two options, but soon you might need three, and it's easier to add a third if you don't change the type later. Anyway, that might be premature optimization since I don't know a concrete third option right now.

Yeah, I agree. Thank you for the suggestion. I'll change so.

**Review by mikeurbach — APPROVED** [review]
I think this part is looking good. @darthscsi any concerns?

File: include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td [hunk 3]
```diff
@@ -338,3 +338,28 @@ def InnerSymbol : OpInterface<"InnerSymbolOpInterface"> {
...
 def InnerSymbolTable : NativeOpTrait<"InnerSymbolTable">;
+
+def FNamableOp : OpInterface<"FNamableOp"> {
+  let cppNamespace = "circt::firrtl";
+  let description = "Provide common connection information.";
+
+  let methods = [
+    InterfaceMethod<"Return a destination of connection.",
+    "bool", "hasDroppableName", (ins), [{}], /*defaultImplementation=*/[{
+      return this->getOperation()->getAttr("nameKind").template
```
Comment (darthscsi) [inline include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td:349]: getAttrOfType

File: include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td [hunk 4]
```diff
@@ -338,3 +338,28 @@ def InnerSymbol : OpInterface<"InnerSymbolOpInterface"> {
...
+    }]>,
+    InterfaceMethod<"Make the name droppable.",
+    "void", "dropName", (ins), [{}], /*defaultImplementation=*/[{
+      this->getOperation()->setAttr("nameKind",
+        NameKindEnumAttr::get(this->getOperation()->getContext(),
+                              NameKindEnum::DroppableName));
+    }]>,
+    InterfaceMethod<"Return the name.",
+    "mlir::StringAttr", "getName", (ins), [{}], /*defaultImplementation=*/[{
+      return this->getOperation()->getAttr("name").template
```
Comment (darthscsi) [inline include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td:361]: getAttrOfType

File: include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td [hunk 5]
```diff
@@ -199,22 +213,24 @@ def NodeOp : ReferableDeclOp<"node",
...
-                   CArg<"StringAttr", "StringAttr()">:$inner_sym), [{
+    OpBuilder<(ins "::mlir::Type":$elementType, "::mlir::Value":$input,
+                  CArg<"StringRef", "{}">:$name,
+                  CArg<"ArrayAttr", "ArrayAttr()">:$annotations,
+                  CArg<"StringAttr", "StringAttr()">:$inner_sym), [{
+      if(!annotations)
+        annotations = $_builder.getArrayAttr({});
       return build($_builder, $_state, elementType, input, name,
-                   $_builder.getArrayAttr(annotations), inner_sym);
+                   annotations, inner_sym, inferNameKind(name));
```
Comment (darthscsi) [inline include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td:228]: I'm not sure this is a good idea for the builders.  infer should only be run by builders called from the parser.  Putting it here encourages random other code to use it.

File: include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td [hunk 6]
```diff
@@ -125,34 +127,35 @@ def GetWidthAsIntAttr : NativeCodeCall<
 
 // node(x,name),!dnt -> node(x, "")
 def DropNameNode : Pat<
-  (NodeOp $x, $name, $annotations, $inner_sym),
-  (NodeOp $x, (GetEmptyString), $annotations, $inner_sym),
+  (NodeOp $x, $name, $annotations, $inner_sym, $nameKind),
+  (NodeOp $x, (GetEmptyString),  $annotations, $inner_sym, (GetDroppableNameAttr)),
   [(uselessNameAttr $name), (NonEmptyAttr $name), (NullAttr $inner_sym)]>;
```
Comment (darthscsi) [inline include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td:132]: rename uselessNameAttr?

Comment (darthscsi) [reply include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td:132]: A also thought the DropName* weren't being used any more.  They probably shouldn't.  We shouldn't drop names anymore.  A droppable name should be kept and moved to expressions.

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp [hunk 7]
```diff
@@ -62,15 +62,16 @@ bool circt::firrtl::isUselessName(StringRef name) {
...
-/// Return true if this is a useless temporary name produced by FIRRTL.  We
-/// drop these as they don't convey semantic meaning.
-bool circt::firrtl::isUselessName(Operation *op) {
-  if (auto wire = dyn_cast<WireOp>(op))
-    return isUselessName(wire.name());
-  if (auto node = dyn_cast<NodeOp>(op))
-    return isUselessName(node.name());
-  if (auto reg = dyn_cast<RegOp>(op))
-    return isUselessName(reg.name());
+NameKindEnum circt::firrtl::inferNameKind(StringRef name) {
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/FIRRTLFolds.cpp:65]: This should be a parser function.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 8]
```diff
@@ -3284,6 +3301,34 @@ static void printElidePortAnnotations(OpAsmPrinter &p, Operation *op,
...
 
+//===----------------------------------------------------------------------===//
+// NameKind Custom Directive
+//===----------------------------------------------------------------------===//
+
+static ParseResult parseNameKind(OpAsmParser &parser,
+                                 firrtl::NameKindEnumAttr &result) {
+  StringRef keyword;
+
+  if (!parser.parseOptionalKeyword(&keyword,
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:3312]: do any of the parsers have the possibility of being confused by something whose name is "interesting_name"?

File: test/Dialect/FIRRTL/canonicalization.mlir [hunk 9]
```diff
@@ -2146,7 +2146,7 @@ firrtl.module @namedrop(in %clock: !firrtl.clock, in %reset: !firrtl.uint<1>, in
   %_T_1 = firrtl.wire : !firrtl.uint<1>
   %_T_2 = firrtl.reg %clock : !firrtl.uint<1>
   %_T_3 = firrtl.regreset %clock, %reset, %c1_ui1 : !firrtl.uint<1>, !firrtl.uint<1>, !firrtl.uint<1>
-  %a = firrtl.mem Undefined {depth = 8 : i64, name = "_T_5", portNames = ["a"], readLatency = 0 : i32, writeLatency = 1 : i32} : !firrtl.bundle<addr: uint<3>, en: uint<1>, clk: clock, data flip: uint<1>>
+  %a = firrtl.mem Undefined  {depth = 8 : i64, name = "_T_5", portNames = ["a"], readLatency = 0 : i32, writeLatency = 1 : i32} : !firrtl.bundle<addr: uint<3>, en: uint<1>, clk: clock, data flip: uint<1>>
```
Comment (darthscsi) [inline test/Dialect/FIRRTL/canonicalization.mlir:2149]: super nit: unnecessary whitespace change.

**Review by darthscsi — APPROVED** [review]
Please narrow down inference so it only appears explicitly in the parser and not the builders.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 8 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:3312]: No, I don't think so. A mnemonic or symbol name can be a precedence of nameKind, but "interesting_name" cannot be mnemonic or symbols.

**Comment by uenoku** [comment]
* Builders are modified to take nameKind explicitly
* `inferNameKind` is moved to parser.
* Added FNambleOp to CHIRTL memory operations

File: include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td (hunk 6 above)
Comment (uenoku) [reply include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td:132]: `uselessNameAttr` is intentionally used because we want to drop names with `_` prefix. I agree that these canonicalizers are not used anymore so I'll do in a follow-up.

**Review by youngar — APPROVED** [review]
sorry for the late review!

**Comment by uenoku** [comment]
Thanks, those should be address by https://github.com/llvm/circt/pull/3312
